### DISCUSSION TITLE
How to Run Notebooks Locally

### DISCUSSION THREAD

**Discussion by srcarter3** [Unidata/pyaos-ams-2021#126] [General]
Just a quick note to everyone.  We highly encourage you to develop your notebooks and code in our [Jupyterhub](https://pyaos-workshop.unidata.ucar.edu/).  However, as we've mentioned a few times, this hub will be taken down permanently on January 19th, and I'm sure at least a few of your are wondering how you can work with the material (the trainings we've provided, as well as what you've created yourself throughout the week) after the hub is gone.
I thought I would start a brief discussion here where we'll post some instructions of how to set up your own computer so you can run your own material after the workshop is over.  Stay tuned for more posts with screenshots and detailed instructions.
Hope the week has started off well for everyone!

**Comment by srcarter3** [comment]
# For Mac Users

## Download and Install Miniconda
1. Go to the [miniconda website](https://docs.conda.io/en/latest/miniconda.html) and download the **Python 3.8** 64-bit "pkg" installer for Mac or click [**here**](https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.pkg) to download it directly.
![Screen Shot 2021-01-11 at 3 37 27 PM](https://user-images.githubusercontent.com/21114460/104246445-413c6e00-5423-11eb-8da6-3d3de5e49db7.png)
2. Once that's downloaded, run the pkg file.  Go through the installation guide and accept their terms/licenses.

    ![Screen Shot 2021-01-11 at 3 40 43 PM](https://user-images.githubusercontent.com/21114460/104246524-67620e00-5423-11eb-9e51-42328c35135f.png)

## Download the Environment.yml
1. After miniconda has been installed successfully, go to our [Jupyter hub](https://pyaos-workshop.unidata.ucar.edu/) and download the "environment.yml" file found in the **pyaos-ams-2021** directory.

    ![Screen Shot 2021-01-11 at 3 44 49 PM](https://user-images.githubusercontent.com/21114460/104246856-f5d68f80-5423-11eb-8182-e6d2b97a865f.png)
2. I would suggest copying or moving this file into a more permanent location on your computer, possibly in your **home directory**.

## Create Conda Environment
1.  Open a new terminal (this can be done by pressing the command key + space, and then then typing "terminal", and then enter:

    ![Screen Shot 2021-01-11 at 3 47 07 PM](https://user-images.githubusercontent.com/21114460/104247057-46e68380-5424-11eb-8043-f312ce7c8f9d.png)
2. If Conda installed and activated properly you should see **(base)** on the left hand side of the prompt in the terminal window:
   >Note: Your terminal may have a different color scheme than mine, that's okay

   ![Screen Shot 2021-01-11 at 3 48 23 PM](https://user-images.githubusercontent.com/21114460/104247166-74cbc800-5424-11eb-8e59-c3bd6802c4d6.png)

   > If you **do not** see (base) in your terminal, go to your home directory:
`cd ~`
 and run the following command:
 `source .bash_profile`

3. Go to the directory where you saved the "environment.yml" file and run the following command:
`conda env create -f environment.yml`

    ![Screen Shot 2021-01-11 at 3 52 01 PM](https://user-images.githubusercontent.com/21114460/104247396-f6bbf100-5424-11eb-91a7-12474a0db79c.png)
4. You should now see a "spinning" icon in the upper right and on the command line.  This process can take a while (~5 minutes or so).  

    ![Screen Shot 2021-01-11 at 3 52 24 PM](https://user-images.githubusercontent.com/21114460/104247481-23700880-5425-11eb-96f8-04b852bf10c6.png)
5. Please wait until it is done to proceed before proceeding with the **Run Jupyter Notebooks Locally** step.  You will know it's done when the command line looks like this:

    ![Screen Shot 2021-01-11 at 4 02 07 PM](https://user-images.githubusercontent.com/21114460/104248251-6383bb00-5426-11eb-81ac-656bc4536ec4.png)


## Download Jupyter Notebooks
1.  While the conda environment is being created you can take this time to go download any and all notebook files from the Jupyterhub.  These can be files you created or files we have put up for training references.  The easiest way to download notebooks is to navigate to their folder, right click the ones you're interested in and select "Download".

    ![Screen Shot 2021-01-11 at 3 55 29 PM](https://user-images.githubusercontent.com/21114460/104247669-73e76600-5425-11eb-8367-7b9f0e46c7f4.png)
2. After those have downloaded, I suggest making a new folder somewhere and copying/moving these notebook files into them.

## Run Jupyter Notebooks Locally
1. Now that you have successfully created your new conda environment, activate it by typing the following into the terminal:
`conda activate pyaos-ams-2021`
2. You will now see **(pyaos-ams-2021)** on the left hand side of the terminal prompt:

    ![Screen Shot 2021-01-11 at 4 03 29 PM](https://user-images.githubusercontent.com/21114460/104248356-90d06900-5426-11eb-81ed-caa5e5676975.png)
3. Navigate (using `cd`) to the directory you saved your notebooks in and run `jupyter notebook`, it will open a browser showing you all the notebooks you have in that directory:

    ![Screen Shot 2021-01-11 at 4 15 14 PM](https://user-images.githubusercontent.com/21114460/104249228-3932fd00-5428-11eb-9f34-c31cec452092.png)
    ![Screen Shot 2021-01-11 at 4 33 38 PM](https://user-images.githubusercontent.com/21114460/104250404-caa36e80-542a-11eb-8a64-bede00e532fb.png)

4. Click on any notebook you want, and it will open a Jupyter editor very similar to the one you see in our JupyterHub.  Run, edit, and save the code however you like!  It's all running entirely on your machine, so no need to access our hub, or even access the internet for that matter. 

    ![Screen Shot 2021-01-11 at 4 16 08 PM](https://user-images.githubusercontent.com/21114460/104249279-5962bc00-5428-11eb-9433-14b5394916f2.png)

## Enjoy!

### Feel free to comment on this discussion and let us know if you have any questions.

**Comment by tiffanycmeyer13** [comment]
# For Windows Users
## Download and Install Miniconda
1. Go to the [miniconda website](https://docs.conda.io/en/latest/miniconda.html) and download the **Python 3.8** 64-bit executable for Windows or click [here](https://repo.anaconda.com/miniconda/Miniconda3-latest-Windows-x86_64.exe) to download it directly.

![image](https://user-images.githubusercontent.com/65614396/104250346-0b06ea80-5433-11eb-8df3-35cd2a59d92d.png)

2. Once that's downlaoded, run the exe file. Go through the installation guide, accept their terms/licenses, and follow the default options:

![image](https://user-images.githubusercontent.com/65614396/104247499-8bc2e800-542d-11eb-97db-1d5bd8fa1ea3.png)

## Download the environment.yml file

1. After miniconda has been installed successfully, go to our [Jupyter hub](https://pyaos-workshop.unidata.ucar.edu/) and download the `environment.yml` file in the **pyaos-ams-2021** directory.

![image](https://user-images.githubusercontent.com/65614396/104250771-ee1ee700-5433-11eb-95c5-4ed9d7364f8f.png)

2. I would suggest copying or moving this file into a more permanent location on your computer, possibly in your **home directory**.

3. Using a text editor open the "environment.yml" file and remove the `   - xesmf>=0.5.0` line. (The xarray_interpolation.ipynb notebook will not work with windows) and verify your environment.yml looks like:

```
 name: pyaos-ams-2021
 channels:
   - conda-forge
 dependencies:
   - python=3.9
   - numpy
   - matplotlib
   - cartopy
   - metpy
   - siphon
   - pandas
   - xarray
   - ipywidgets
   - ffmpeg
   - boto3
   - jupyterlab
   - python-awips
   - xskillscore
   - cfgrib
   - netcdf4
   # Fix for tab completion
   - jedi=0.17.2
```
## Create Conda Environment
1. Open an anaconda command prompt. From the Windows start menu open Anaconda3 --> Anaconda Powershell Prompt

![image](https://user-images.githubusercontent.com/65614396/104254540-ae103200-543c-11eb-9185-ab1cb79d2fc4.png)

2. If Conda installed and activated properly you should see (base) on the left hand side of the prompt in the terminal window:

![image](https://user-images.githubusercontent.com/65614396/104253097-4c9a9400-5439-11eb-99b9-2f4db8f0e9fe.png)

3. Go to the directory where you saved  the "environment.yml" file and run the following command:
`conda env create -f environment.yml`

![image](https://user-images.githubusercontent.com/65614396/104253138-676d0880-5439-11eb-9ec9-58e53d66c0f6.png)

4. You should now see the output in the terminal as packages are installed. This process can take a while (~5 minutes or so).

![image](https://user-images.githubusercontent.com/65614396/104252977-02b1ae00-5439-11eb-8d83-e6366f4bf576.png)

5. Please wait until it is done to proceed before proceeding with the Running Jupyter Notebooks step. You will know it's done when the command line looks like this:

![image](https://user-images.githubusercontent.com/65614396/104253732-d4cd6900-543a-11eb-9ee3-642eba6b484f.png)

## Downloading Jupyter Notebooks

1. While the conda environment is being created you can take this time to go download any and all notebook files from the Jupyterhub. These can be files you created or files we have put up for training references. The easiest way to download notebooks is to navigate to their folder, right click the ones you're interested in and select "download".

![image](https://user-images.githubusercontent.com/65614396/104252250-35f33d80-5437-11eb-810b-b69beb3036eb.png)

2. After those have downloaded, I suggest making a new folder somewhere and copying/moving these notebook files into them.

## Running Jupyter Notebooks

1. Now that you have successfully created your new conda environment, activate it by typing the following into the anaconda powershell prompt:
 `conda activate pyaos-ams-2021`

2. You will now see (**pyaos-ams-2021**) on the left hand side of the terminal prompt:

![image](https://user-images.githubusercontent.com/65614396/104253312-c7fc4580-5439-11eb-803b-e08d98c12b79.png)

3. Navigate (using `cd`) to the directory you saved your notebooks in and run `jupyter notebook`, it will open a browser showing you all the notebooks you have in that directory:

![image](https://user-images.githubusercontent.com/65614396/104253460-16a9df80-543a-11eb-823e-da75693b8f3f.png)

4. Click on any notebook you want, and it will open a Jupyter editor very similar to the one you see in our JupyterHub. Run, edit, and save the code however you like! It's all running entirely on your machine, so no need to access our hub, or even access the internet for that matter.

![image](https://user-images.githubusercontent.com/65614396/104253643-9f288000-543a-11eb-8483-452aac81e0e9.png)

## Enjoy!
